web/src/components/storage/dasd/FormatActionHandler.tsx

```diff
@@ -0,0 +1,207 @@
+import Popup from "~/components/core/Popup";
+import { DASDDevice } from "~/types/dasd";
+import { useFormatDASDMutation } from "~/queries/storage/dasd";
+import { sprintf } from "sprintf-js";
+import { _ } from "~/i18n";
+import { isEmpty } from "radashi";
+
+/**
+ * Shared type for defining props used by all DASD format-related dialogs and
+ * the main controller component.
+ */
+type CommonFormatDASDProps = {
+  /** A single DASD device, used for single-device dialogs. */
+  device: DASDDevice;
+  /** An array of DASD devices selected for formatting. */
+  devices: DASDDevice[];
+  /** Callback triggered when the user confirms the format operation. */
+  onCancel?: () => void;
+  /** Callback triggered when the user cancels the operation. */
+  onAccept?: () => void;
+};
+
+/**
+ * Renders ids of given DASD devices in a simple, unordered list
+ */
+const DevicesList = ({ devices }: Pick<CommonFormatDASDProps, "devices">) => (
+  <List>
+    {devices.map((d: DASDDevice) => (
+      <ListItem key={d.id}>{d.id}</ListItem>
+    ))}
+  </List>
+);
+
+/**
+ * Renders a popup indicating a specific device is offline.
+ * Used when attempting to format a single device that is disabled.
+ */
+const DeviceOffline = ({
+  device,
+  onCancel,
+}: Pick<CommonFormatDASDProps, "device" | "onCancel">) => {
+  return (
+    <Popup isOpen title={sprintf(_("Cannot format %s"), device.id)}>
+      <Stack hasGutter>
+        <Content>{_("It is offline and must be activated before formatting it.")}</Content>
+      </Stack>
+      <Popup.Actions>
+        <Popup.Confirm onClick={onCancel}>{_("Accept")}</Popup.Confirm>
+      </Popup.Actions>
+    </Popup>
+  );
+};
+
+/**
+ * Shows a popup listing multiple offline devices,
+ * preventing a format action on them.
+ */
+const SomeDevicesOffline = ({
+  devices,
+  onCancel,
+}: Pick<CommonFormatDASDProps, "devices" | "onCancel">) => {
+  const offlineDevices = devices.filter((d) => !d.enabled);
+  const totalOffline = offlineDevices.length;
+
+  return (
+    <Popup isOpen title={_("Cannot format all the selected devices")}>
+      <Stack hasGutter>
+        <Content>
+          {sprintf(_("Below %s devices are offline and cannot be formatted."), totalOffline)}
+        </Content>
+        <Content>{_("Unselect or activate them and try it again.")}</Content>
+        <DevicesList devices={offlineDevices} />
+      </Stack>
+      <Popup.Actions>
+        <Popup.Confirm onClick={onCancel}>{_("Accept")}</Popup.Confirm>
+      </Popup.Actions>
+    </Popup>
+  );
+};
+
+/**
+ * Renders a format confirmation dialog for formatting a single device.
+ */
+const DeviceFormatConfirmation = ({
+  device,
+  onAccept,
+  onCancel,
+}: Pick<CommonFormatDASDProps, "device" | "onAccept" | "onCancel">) => {
+  return (
+    <Popup isOpen title={sprintf(_("Format device %s"), device.id)}>
+      <Content>
+        <Stack hasGutter>
+          <Text isBold>{_("This action could destroy any data stored on the device.")}</Text>
```

Comment:
This is the original message was there before starting this refactorization, but I'd say it is "will" not "could".